### PR TITLE
Continue indexing on deterministic error

### PR DESCRIPTION
**Pull request by leoyvens**
This implements the non-fatal errors proposal, by which a subgraph which has synced will continue indexing when encountering a deterministic error, skipping the handler and discarding any changes done while processing it. The errors will be registered when transacting operations for a block, and will be reverted if necessary. To enable non-fatal errors the manifest must include the section:
```
features:
  - nonFatalErrors
```

By default data is hidden from queries in the presence of indexing errors (except for the `_meta` field). To receive data in a query the `subgraphError: allow` argument must be passed in. See the `non_fatal_errors` test for examples of what the responses look like. The `_meta` field gains an `hasIndexingErorrs: Boolean!` field as a way of checking for errors. Additionally if indexing errors are present a graphql error with the message `"indexing_error"` will always be added.

This is ready for review. The `nonFatalErrors` field of the indexing status API will be implemented in a follow up. I'd also like to add an integration test before merging this.

### CODE REVIEW

File: store/postgres/src/metadata.rs [hunk 1]
```diff
@@ -883,3 +886,21 @@ fn check_health(conn: &PgConnection, id: &SubgraphDeploymentId) -> Result<(), an
...
+    conn: &PgConnection,
+    id: &SubgraphDeploymentId,
+    reverted_block: BlockNumber,
+) -> Result<(), anyhow::Error> {
+    use subgraph_error as e;
+
+    delete(
+        e::table
+            .filter(e::subgraph_id.eq(id.as_str()))
+            .filter(sql("lower(block_range) >= $1").bind::<Integer, _>(reverted_block)),
```
Comment (lutter) [inline store/postgres/src/metadata.rs]: this should use `block_range::BLOCK_RANGE_COLUMN` as the name of the block range column

File: graph/src/components/store.rs [hunk 2]
```diff
@@ -1282,6 +1282,7 @@ pub trait EthereumCallCache: Send + Sync + 'static {
 }
 
 /// Store operations used when serving queries
+#[async_trait]
```
Comment (lutter) [inline graph/src/components/store.rs:1285]: Similarly, let's please hold off on asyncifying store operations.

File: graph/src/data/query/error.rs [hunk 3]
```diff
@@ -252,6 +252,12 @@ impl From<StoreError> for QueryExecutionError {
     }
 }
 
+impl From<anyhow::Error> for QueryExecutionError {
```
Comment (lutter) [inline graph/src/data/query/error.rs:255]: It seems weird that any `anyhow::Error` gets translated into a `QueryExecutionError::StoreError`

File: graphql/src/query/ext.rs [hunk 4]
```diff
@@ -68,64 +69,22 @@ impl Default for BlockConstraint {
...
-                    Err(invalid_argument("block", self, value))
-                }
-            } else {
-                Err(invalid_argument("block", self, value))
-            }
+        if let Some(hash) = map.get("hash") {
+            Ok(BlockConstraint::Hash(TryFromValue::try_from_value(hash)?))
+        } else if let Some(number_value) = map.get("number") {
+            let number: u64 = TryFromValue::try_from_value(number_value)?;
+            Ok(BlockConstraint::Number(TryFrom::try_from(number)?))
```
Comment (lutter) [inline graphql/src/query/ext.rs:85]: Nice! No idea what drugs I was on when I initially wrote that code.

File: graphql/src/schema/api.rs [hunk 5]
```diff
@@ -174,6 +216,26 @@ fn add_block_height_type(schema: &mut Document) {
     schema.definitions.push(def);
 }
 
+fn add_subgraph_error_policy_type(schema: &mut Document) {
```
Comment (lutter) [inline graphql/src/schema/api.rs:219]: Not sure if you noticed, but I added a `meta.graphql` file a while ago. Would be cleaner to add this type definition to that file, too, since a GraphQL schema definition is much easier to read than Rust code that builds the AST.

File: graphql/tests/query.rs [hunk 6]
```diff
@@ -1637,3 +1640,27 @@ fn can_query_meta() {
...
+            deterministic: true,
+        };
+
+        transact_errors(&*STORE, id.clone(), BLOCK_TWO.block_ptr(), vec![err]).unwrap();
+
+        let query = "query { musician(id: \"m1\") { id } }";
+        let query = graphql_parser::parse_query(query).unwrap();
+
+        dbg!(execute_query_document(&id, query).await);
+        assert!(false);
```
Comment (lutter) [inline graphql/tests/query.rs]: Looks like this test needs just a little more work ;)

File: graph/src/data/query/result.rs [hunk 7]
```diff
@@ -97,54 +172,65 @@ impl QueryResult {
...
+        serialize_with = "serialize_data"
+    )]
+    data: Option<Data>,
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    errors: Vec<QueryError>,
+    #[serde(
+        skip_serializing_if = "Option::is_none",
+        serialize_with = "serialize_extensions"
+    )]
+    extensions: Option<q::Value>,
```
Comment (lutter) [inline graph/src/data/query/result.rs:191]: We should really nuke `extensions` since it is always `None`

**Review by lutter — CHANGES_REQUESTED** [review]

File: store/postgres/src/metadata.rs [hunk 8]
```diff
@@ -751,6 +753,62 @@ pub fn reassign_subgraph(
...
+            e::subgraph_id.eq(subgraph_id.as_str()),
+            e::message.eq(message),
+            e::handler.eq(handler),
+            e::deterministic.eq(deterministic),
+            e::block_hash.eq(block_ptr.as_ref().map(|ptr| ptr.hash.as_bytes())),
+            e::block_number.eq(sql(&format!("{}::numeric", block_num))),
+            e::block_range.eq((Bound::Included(block_num), Bound::Unbounded)),
+        ))
+        .on_conflict_do_nothing()
+        .execute(conn)?;
```
Comment (lutter) [inline store/postgres/src/metadata.rs:795]: You also need to clamp the range of precious errors, otherwise, if you insert an error with range `[1,)`, you will never be able to insert another error for that subgraph.

And since `block_number = lower(block_range)`, do we even need that column?

**Review by lutter — CHANGES_REQUESTED** [review]

File: store/postgres/src/metadata.rs (hunk 8 above)
Comment (lutter) [reply store/postgres/src/metadata.rs:795]: Since `on_conflict_do_nothing()` can mask problems silently, we need tests that check the various combinations of storing errors and that they work; at a minimum, it should check
* storing the same error twice stores only one error
* storing the same error at two different blocks stores two errors
* reverting correctly reverts errors

And it would be really nice if errors had a `created_at` column to make it easier to find new errors.

File: graph/src/components/store.rs [hunk 9]
```diff
@@ -1448,6 +1451,19 @@ impl EntityCache {
         Ok(())
     }
 
+    /// A copy of the pending updates. Those are kept in a persistent data structure so this is a
+    /// cheap operation.
+    pub(crate) fn updates_snapshot(&self) -> im::HashMap<EntityKey, Option<Entity>> {
+        self.updates.clone()
+    }
+
+    pub(crate) fn restore_updates_snapshot(
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply graph/src/components/store.rs]: Good news is I implemented this without the persistent data structure, which really was bad for performance. Bad news is the code touched some tricky logic in the entity cache and is not super robust since it requires explicitly entering and exiting the handler. But I added and end-to-end test and plenty of asserts.

File: core/src/subgraph/instance_manager.rs [hunk 10]
```diff
@@ -662,7 +665,25 @@ where
     )
     .await
     {
+        // The triggers were processed but some were skipped due to deterministic errors.
+        Ok(block_state) if block_state.has_errors() => {
+            use SubgraphFeatures::*;
+
+            // TODO: || subgraph still syncing
+            if !ctx.inputs.features.contains(&nonFatalErrors) {
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs]: Changed.

File: runtime/wasm/src/module/mod.rs [hunk 11]
```diff
@@ -245,9 +245,9 @@ impl WasmInstance {
         };
 
         if let Some(deterministic_error) = deterministic_error {
-            // Log the error and restore the updates snaphsot, effectively skipping the handler.
+            // Log the error and restore the updates snaphsot, effectively reverting the handler.
             error!(&self.instance_ctx().ctx.logger,
-                "Handler skipped due to execution failure";
+                "Handler reverted";
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply runtime/wasm/src/module/mod.rs]: Changed to the previous message.

File: core/src/subgraph/instance_manager.rs [hunk 12]
```diff
@@ -25,6 +25,10 @@ lazy_static! {
             .unwrap_or("10000".into())
             .parse::<usize>()
             .expect("invalid GRAPH_ENTITY_CACHE_SIZE");
+
+    // Keep deterministic errors non-fatal even if the subgraph is pending. For test purposes.
+    pub static ref DISABLE_FAIL_FAST: bool =
+        std::env::var("GRAPH_DISABLE_FAIL_FAST").is_ok();
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply core/src/subgraph/instance_manager.rs:35]: Yhea graph node itself, updated the comment.

File: store/postgres/src/network_store.rs [hunk 13]
```diff
@@ -48,6 +51,7 @@ impl NetworkStore {
     }
 }
 
+#[async_trait::async_trait]
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply store/postgres/src/network_store.rs:55]: We agreed that this is ok.

File: store/postgres/src/network_store.rs [hunk 14]
```diff
@@ -169,6 +173,14 @@ impl StoreTrait for NetworkStore {
         self.store.deployment_state_from_id(id)
     }
 
+    async fn fail_subgraph(
+        &self,
+        id: SubgraphDeploymentId,
+        error: SubgraphError,
+    ) -> Result<(), anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (leoyvens) [reply store/postgres/src/network_store.rs]: Yes it wasn't a great idea to use anyhow, I switched this and a bunch of other places to StoreError.

File: store/postgres/src/metadata.rs (hunk 1 above)
Comment (leoyvens) [reply store/postgres/src/metadata.rs]: Done.

File: graph/src/data/query/error.rs (hunk 3 above)
Comment (leoyvens) [reply graph/src/data/query/error.rs]: Super weird, refactored this away.

File: graphql/src/schema/api.rs (hunk 5 above)
Comment (leoyvens) [reply graphql/src/schema/api.rs]: That is nice! Moved it there.

File: graph/src/data/query/result.rs (hunk 7 above)
Comment (leoyvens) [reply graph/src/data/query/result.rs]: Yay done

File: store/postgres/src/metadata.rs (hunk 8 above)
Comment (leoyvens) [reply store/postgres/src/metadata.rs:793]: Dropped `block_number`, added `created_at`, added some tests that the conflict do nothing works.

**Comment by leoyvens** [comment]
@lutter I believe I've addressed all your comments, happy reviewing and rebasing!

File: graph/src/data/query/error.rs [hunk 15]
```diff
@@ -252,11 +252,11 @@ impl From<StoreError> for QueryExecutionError {
...
-impl From<anyhow::Error> for QueryExecutionError {
-    fn from(e: anyhow::Error) -> Self {
-        QueryExecutionError::StoreError(CloneableFailureError(Arc::new(e.compat_err())))
-    }
-}
+// impl From<anyhow::Error> for QueryExecutionError {
+//     fn from(e: anyhow::Error) -> Self {
+//         QueryExecutionError::StoreError(CloneableFailureError(Arc::new(e.compat_err())))
+//     }
+// }
```
Comment (lutter) [inline graph/src/data/query/error.rs]: Leftover from trying things out? Should probably be removed

File: store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql [hunk 16]
```diff
@@ -0,0 +1,9 @@
+alter table
+    subgraphs.subgraph_error
+drop
+    column block_number;
+
+alter table
+    subgraphs.subgraph_error
+add
+    column created_at timestamp NOT NULL DEFAULT (now() at time zone 'utc');
```
Comment (lutter) [inline store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql:9]: The data type should be `timstamptz` so that [the timezone gets stored](https://www.postgresql.org/docs/9.6/datatype-datetime.html), too. Just `timestamp` defaults to `without time zone`. And I think you can just say `default now()` if we use a type with time zone.

File: store/postgres/src/metadata.rs [hunk 17]
```diff
@@ -914,3 +914,66 @@ pub(crate) fn revert_subgraph_errors(
 
     check_health(conn, id)
 }
+
+#[test]
```
Comment (lutter) [inline store/postgres/src/metadata.rs:918]: I am not against this, but I thought we put tests that require external components like the store into `tests/`?

File: store/postgres/src/entities.rs [hunk 18]
```diff
@@ -773,6 +773,7 @@ pub fn delete_all_entities_for_test_use_only(
         delete from subgraphs.ethereum_contract_data_source_template;
         delete from subgraphs.ethereum_contract_data_source_template_source;
         delete from subgraphs.ethereum_contract_event_handler;
+        delete from subgraphs.subgraph_error;
```
Comment (lutter) [inline store/postgres/src/entities.rs:776]: Is this necessary? There's an `on delete cascade` on the foreign key `subgraph_error.subgraph_id`

File: store/postgres/src/metadata.rs (hunk 17 above)
Comment (lutter) [reply store/postgres/src/metadata.rs:918]: Actually, leave that here (fear of merge conflicts) - I can always move it later

**Review by lutter — APPROVED** [review]
Great! Once the tests pass, this is ready to merge.

File: store/postgres/src/entities.rs (hunk 18 above)
Comment (leoyvens) [reply store/postgres/src/entities.rs]: Oh that's why we could get away without it. Removed it.

File: store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql (hunk 16 above)
Comment (leoyvens) [reply store/postgres/migrations/2020-12-7-190800_subgraph_error_changes/up.sql]: Done.
